text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,403 @@
+package, and ship dependencies. At this point in time, we are not advocating
+for the deprecation or removal of the present way of managing dependencies,
+however, we hope over time that will be the natural evolution of things.
+
+## Motivation
+Presently, a Paketo buildpack that has binary dependencies will list metadata
+about these dependencies within its `buildpack.toml` file. This includes a URL
+from which the dependency can be downloaded, and also a checksum hash and other
+metadata like PURL and CPEs.
+
+Libraries like `libpak` and `packit` then provide convenience methods for
+fetching dependencies using this metadata, verifying the download, and caching
+the download result. In addition, they provide tooling to download and store
+these dependencies within buildpack images for distribution in offline
+environments.
+
+There are also tools published by the project to manage the entries within
+`buildpack.toml` through CI pipelines so that dependencies metadata is kept
+up-to-date with upstream sources. Unfortunately, this represents a large amount
+of toil for the buildpacks team.
+
+As an example of the toil mentioned in a language family like Java, there are
+daily project dependencies that need to be updated. This requires reviewing and
+merging PRs into the buildpacks to adjust `buildpack.toml` dependency metadata.
+Once PRs are merged, a component buildpack needs to be released, followed by a
+composite buildpack and then a builder release. This is because most users
+don’t consume buildpacks directly, they consume builders which include
+buildpacks.
+
+This all has to be done as aggressively as possible so that we are shipping
+dependencies, in particular those with security fixes, quickly. This is because
+with metadata in `buildpack.toml`, even if an upstream project releases a bug
+or security fix, buildpack users cannot get that fix until we update and
+release component and composite buildpacks as well as the builder.
+
+There is also toil associated with the tools and pipelines used for this
+process. The tools have bugs and need to be updated. At present, the tools we
+use to manage all of these updates do not scale well either. In particular
+Github Actions, we have had a number of issues hitting rate limits and usage
+caps. This gets worse when there are a lot of dependencies to watch, for
+example, if your buildpack has multiple version lines or different sets of
+packages for a dependency.
+
+The whole process puts an additional maintenance burden on the project
+maintainers and project resources. This is not the type of work that a casual
+contributor to Paketo will do and as we add more dependencies the burden only
+increases on the maintainer teams.
+
+The motivation of this proposal is to…
+
+- Reduce the burden and toil for Paketo buildpack maintainer teams
+- Continue publishing dependency updates in a timely and secure manner
+- Decouple installing dependencies from configuring them
```

Comment:
This goal makes sense, but I don't think this RFC expands on how we would accomplish this.

Reply:
I think there is a bit of deliberate ambiguity on this point. The RFC is to split out dependencies so they are separate, once that happens there is freedom to do more. This is an example of something this freedom could enable.

A buildpack could install and configure a dependency or you could break things up and have one buildpack that installs many dependencies and others that configure or run dependencies. I think technically you could do something like this with metadata included in a buildpack but in practice that would get cumbersome.

Reply:
Ok, that seems fair enough. I can see how making the metadata dynamic gives flexibility, and we don't need to be prescriptive in this RFC.

I think Paketo as an OSS project should probably coalesce on a single "best practice" for the buildpacks we maintain. I'd love to use this as an opportunity to bring the packit/libpak buildpacks closer together, rather than see them drift further.

While we don't have to put this in the RFC, do you have an opinion on what structure you'd like to see for the libpak buildpacks? I'm assuming we would start with one buildpack per language-family that provides the metadata in the default location, and another buildpack per language-family that provides the dependencies in the default location. I would be perfectly happy combining those two into a single buildpack, though. Does that align with your expectations, @dmikusa ?

Reply:
> While we don't have to put this in the RFC, do you have an opinion on what structure you'd like to see for the libpak buildpacks?

The structure of the metadata should be consistent, it's described in this RFC. How we get the metadata shuffled around should be something we agree upon as a project so it's consistent.

> I'm assuming we would start with one buildpack per language-family that provides the metadata in the default location, and another buildpack per language-family that provides the dependencies in the default location. I would be perfectly happy combining those two into a single buildpack, though. Does that align with your expectations, @dmikusa ?

At a high level, my thought was that we'd have a buildpack to distribute the metadata. The metadata would then reference things directly from the Internet, similar to how we do that now. There are still a lot of open questions about how that would work exactly. I think that we should probably have an RFC to discuss that & agree upon things so that we're working consistently across buildpacks. 

I just want it to be separate so that we don't make this RFC any larger than it already is.

Reply:
> The structure of the metadata should be consistent, it's described in this RFC. How we get the metadata shuffled around should be something we agree upon as a project so it's consistent.

💯 

> At a high level, my thought was that we'd have a buildpack to distribute the metadata. The metadata would then reference things directly from the Internet, similar to how we do that now. There are still a lot of open questions about how that would work exactly. I think that we should probably have an RFC to discuss that & agree upon things so that we're working consistently across buildpacks.
>
> I just want it to be separate so that we don't make this RFC any larger than it already is.

Sure, I think that's perfectly reasonable. I'm happy to defer that conversation to a separate RFC, with the understanding that implementing this RFC would be blocked on that downstream RFC.